pkg/curl/schemes.go

```diff
@@ -147,6 +155,18 @@ func (s Schemes) Fetch(ctx context.Context, u *url.URL) (File, error) {
 	return &file{ReaderAt: r, url: u}, nil
 }
 
+func (s Schemes) FetchWithNoCache(ctx context.Context, u *url.URL) (File, error) {
```

Comment:
Hey,

I think it would be good for us to consider an alternative. First of all, you've done really good work here. Finding these memory issues is awesome.

However, the way we are going about fixing it is only going to help our internal use case. Both external users and internal PXE netboot or diskboot use cases are going to run into the same issue, since they all use `Fetch`.

IMO, I think we should make the caching reader more efficient. In fact, let's kill 2 birds with one stone:

- First, the caching reader is the cause of OOM because it needs 2xFileSize memory.
- Later, copying the file from Go to a file in [this function before kexec](https://github.com/u-root/u-root/blob/236bcb5a057c69b708f1844e93d97e248a103c05/pkg/boot/linux.go#L59) is an issue because yet again you need 2xFileSize memory -- the Go buffer holding the kernel won't be freed until the full file is copied into tmpfs.

What if the caching reader just copied the file to a tmpfs in the first place, and there was a special function to get a kexec-compliant file from it (I know this doesn't satisfy the OOM from the internal code -- let's talk about that tomorrow, I also have thoughts.)

```go
// TmpfsFile is not much more than a wrapper around an *os.File guaranteed to be in a tmpfs.
// 
// TmpfsFile only implements io.ReaderAt so that users do not have to deal
// with seeking to the beginning of the file, and so the file can be used by
// multiple goroutines in parallel without worrying about corrupting state.
//
// TmpfsFile does not allow direct access to f in order to satisfy the precondition
// to kexec_file_load that no other open writable FD to this file exist.
type TmpfsFile struct {
  f *os.File
}

func TmpfsFileFrom(r io.Reader) (*TmpfsFile, error) {
  // If available and if it works with kexec_file_load, we should use memfd_create(2) on Linux to make the file
  // private to this process, so nobody can tamper with it between reading, verifying contents, and using it for kexec.
  //
  // It'd be: memfd_create, write to the file, then $roFD = open(/proc/self/fd/$FD, O_RDONLY), close($fd), return $roFD for kexec.
  f, err := ioutil.TempFile("/tmp", "cached-file-")
  if err != nil {
    return nil, err
  }
  defer f.Close()

  // If necessary, use io.CopyBuffer to supply a fixed size buffer for chunked copies.
  if _, err := io.Copy(f, r); err != nil {
    return nil, err
  }
  if err := f.Sync(); err != nil {
    return nil, err
  }

  readOnlyF, err := os.Open(f.Name())
  if err != nil {
    return nil, err
  }
  return &TmpfsFile{f: readOnlyF}, nil
}

// Reader returns a reader seeked to the beginning of the file.
func (tf *TmpfsFile) Reader() io.Reader {
  return uio.Reader(tf.f)
}

func (tf *TmpfsFile) ReadAt(p []byte, off int64) (n int, err error) {
  return tf.f.ReadAt(p, off)
}

// KexecFile returns a file that fulfills the preconditions of an FD passed to kexec_file_load:
//
// (1) must not be opened writable by anything, similar to how you can't run a process if it's opened writable by anything (https://elixir.bootlin.com/linux/v5.12.13/source/fs/kernel_read_file.c#L52 `deny_write_access`)
// (2) must be a regular file. (https://elixir.bootlin.com/linux/v5.12.13/source/fs/kernel_read_file.c#L48)
func (tf *TmpfsFile) KexecFile() *os.File {
  return tf.f
}
```

and in `pkg/boot/linux.go`:

```go
// KexecFiler is implemented by files that can be directly used by kexec_file_load.
type KexecFiler interface {
	// KexecFile returns a file that fulfills the preconditions of an FD passed to kexec_file_load:
	//
	// (1) must not be opened writable by anything, similar to how you can't run a process if
	// it's opened writable by anything
	// (https://elixir.bootlin.com/linux/v5.12.13/source/fs/kernel_read_file.c#L52 `deny_write_access`)
	// (2) must be a regular file. (https://elixir.bootlin.com/linux/v5.12.13/source/fs/kernel_read_file.c#L48)
	KexecFile() *os.File
}

func getKexecFile(r io.ReaderAt, verbose bool) (*os.File, error) {
	if kf, ok := r.(KexecFiler); ok {
		return kf.KexecFile(), nil
	}

	// All that other stuff from LinuxImage.Load
	kernel := uio.Reader(util.TryGzipFilter(r))
	if verbose {
		// In verbose mode, print a dot every 5MiB. It is not pretty,
		// but it at least proves the files are still downloading.
		progress := func(r io.Reader) io.Reader {
			return &uio.ProgressReadCloser{
				RC:       ioutil.NopCloser(r),
				Symbol:   ".",
				Interval: 5 * 1024 * 1024,
				W:        os.Stdout,
			}
		}
		kernel = progress(kernel)
		initrd = progress(initrd)
	}

	// It seams inefficient to always copy, in particular when the reader
	// is an io.File but that's not sufficient, os.File could be a socket,
	// a pipe or some other strange thing. Also kexec_file_load will fail
	// (similar to execve) if anything as the file opened for writing.
	// That's unfortunately something we can't guarantee here - unless we
	// make a copy of the file and dump it somewhere.
	k, err := copyToFile(kernel)
	if err != nil {
		return nil, err
	}
	return k, nil
}

func (li *LinuxImage) Load(verbose bool) error {
	kernel, err := getKexecFile(li.Kernel)
	if err != nil {
		return err
	}
	defer kernel.Close()

	// same with initramfs etc

	return kexec.FileLoad(kernel, ...)
}
```

Reply:
Going to say -- I like the memfd idea more and more. kexec is only supported on Linux for now anyway, and it assures us that nothing else can tamper with the file.

Reply:
It would be helpful to try this out on a 4G setup. After some reading, it seems you would also need to apply appropriate seals in appropriate places per design https://lwn.net/Articles/591108/. 

From reading, my understanding of it is that it would make it cached contents not part of golang heap, but dedicate that into tmpfs. So it does not reduce consumption. But we want cache for Fetch, so that is not wrong.  Since we will read in small chunks, so we limit large go heap allocations. I tend to believe that this will work around the first OOM issue we hit internally. 

**Another thing worth noting** is: initramfs cpio is wrapped at a later stage at boot entry preparation stage, not directly downloaded. So this need to made work with that. I imagine it not clean to deal with.

It also reads like you would apply seals to the tmpfs files so that it can stay intact up until kexec. Then this become a fixed overhead, compression against initramfs to save additional space for kexec would be infeasible. That is not a problem if compression is not needed. Realistically, I found I always need compression to save additional 0.5G space for the flow to work.

Reply:
I like the idea of `TmpfsFile` in general, but what calls `TmpfsFileFrom`? Could this logic go in `copyToFile`?

Also, I am not too clear on the purpose of memfd? Who else would be writing to the file in tmpfs? Seems like it is solving a separate problem.

Reply:
I'm not too happy with this part however:

```
	if kf, ok := r.(KexecFiler); ok {
		return kf.KexecFile(), nil
	}
```

It feels like this adds an "implicit" API between the two packages which will be very difficult to rediscover in the future.

Reply:
Done - discussions done from different threads internally, resolve this now.